File: 2026/regional-andaluza/K/code/dinic.py
from collections import deque

class Dinic:
    def __init__(self, n):
        self.n = n
        self.g = [[] for _ in range(n)]

    def add_edge(self, u, v, c):
        self.g[u].append([v, c, len(self.g[v])])
        self.g[v].append([u, 0, len(self.g[u]) - 1])

    def bfs(self, s, t):
        self.level = [-1] * self.n
        q = deque([s])
        self.level[s] = 0

        while q:
            u = q.popleft()
            for v, cap, rev in self.g[u]:
                if cap > 0 and self.level[v] == -1:
                    self.level[v] = self.level[u] + 1
                    q.append(v)

        return self.level[t] != -1

    def dfs(self, u, t, f):
        if u == t:
            return f

        while self.it[u] < len(self.g[u]):
            i = self.it[u]
            self.it[u] += 1

            v, cap, rev = self.g[u][i]
            if cap > 0 and self.level[v] == self.level[u] + 1:
                pushed = self.dfs(v, t, min(f, cap))
                if pushed > 0:
                    self.g[u][i][1] -= pushed
                    self.g[v][rev][1] += pushed
                    return pushed

        return 0

    def max_flow(self, s, t):
        flow = 0
        INF = 10**9

        while self.bfs(s, t):
            self.it = [0] * self.n
            while True:
                pushed = self.dfs(s, t, INF)
                if pushed == 0:
                    break
                flow += pushed

        return flow


def solve():
    T = int(input())

    for _ in range(T):
        N, M, K = map(int, input().split())
        roles = input().split()
        role_id = {name: i for i, name in enumerate(roles)}

        limit_role = N // K

        S = 0
        first_player = 1
        first_role = 1 + M
        Tnode = 1 + M + K

        dinic = Dinic(Tnode + 1)

        for i in range(M):
            dinic.add_edge(S, first_player + i, 1)

        for i in range(M):
            a, b = input().split()
            dinic.add_edge(first_player + i, first_role + role_id[a], 1)
            dinic.add_edge(first_player + i, first_role + role_id[b], 1)

        for j in range(K):
            dinic.add_edge(first_role + j, Tnode, limit_role)

        flow = dinic.max_flow(S, Tnode)
        print("YES" if flow == M else "NO")


solve()

File: 2026/regional-andaluza/K/code/dinic.cpp
#include <bits/stdc++.h>
using namespace std;

struct Edge {
    int to, cap, rev;
};

struct Dinic {
    int n;
    vector<vector<Edge>> g;
    vector<int> level, it;

    Dinic(int n) : n(n), g(n), level(n), it(n) {}

    void addEdge(int u, int v, int cap) {
        g[u].push_back({v, cap, (int)g[v].size()});
        g[v].push_back({u, 0, (int)g[u].size() - 1});
    }

    bool bfs(int s, int t) {
        fill(level.begin(), level.end(), -1);
        queue<int> q;
        level[s] = 0;
        q.push(s);

        while (!q.empty()) {
            int u = q.front();
            q.pop();

            for (const auto &e : g[u]) {
                if (e.cap > 0 && level[e.to] == -1) {
                    level[e.to] = level[u] + 1;
                    q.push(e.to);
                }
            }
        }

        return level[t] != -1;
    }

    int dfs(int u, int t, int f) {
        if (u == t) return f;

        for (int &i = it[u]; i < (int)g[u].size(); i++) {
            Edge &e = g[u][i];
            if (e.cap > 0 && level[e.to] == level[u] + 1) {
                int pushed = dfs(e.to, t, min(f, e.cap));
                if (pushed > 0) {
                    e.cap -= pushed;
                    g[e.to][e.rev].cap += pushed;
                    return pushed;
                }
            }
        }

        return 0;
    }

    int maxFlow(int s, int t) {
        int flow = 0;
        const int INF = 1e9;

        while (bfs(s, t)) {
            fill(it.begin(), it.end(), 0);
            while (true) {
                int pushed = dfs(s, t, INF);
                if (pushed == 0) break;
                flow += pushed;
            }
        }

        return flow;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    while (T--) {
        int N, M, K;
        cin >> N >> M >> K;

        unordered_map<string, int> roleId;
        for (int i = 0; i < K; i++) {
            string role;
            cin >> role;
            roleId[role] = i;
        }

        int limit = N / K;

        int S = 0;
        int firstPlayer = 1;
        int firstRole = 1 + M;
        int Tnode = 1 + M + K;

        Dinic dinic(Tnode + 1);

        for (int i = 0; i < M; i++) {
            dinic.addEdge(S, firstPlayer + i, 1);
        }

        for (int i = 0; i < M; i++) {
            string a, b;
            cin >> a >> b;
            dinic.addEdge(firstPlayer + i, firstRole + roleId[a], 1);
            dinic.addEdge(firstPlayer + i, firstRole + roleId[b], 1);
        }

        for (int j = 0; j < K; j++) {
            dinic.addEdge(firstRole + j, Tnode, limit);
        }

        int flow = dinic.maxFlow(S, Tnode);

        cout << (flow == M ? "YES" : "NO") << '\n';
    }

    return 0;
}